tools/distpartitioning/globalids.py

```diff
@@ -0,0 +1,279 @@
+    total_nodes = np.sum(sizes)
+    if (total_nodes == 0):
+        print('Rank: ', rank, ' -- All mappings are present locally... No need for to send any info.')
+        return None
+
+    #determine the no. of global_node_ids to send and receive and perform alltoall
+    send_counts = list(torch.Tensor(sizes).type(dtype=torch.int64).chunk(world_size))
+    recv_counts = list(torch.zeros([world_size], dtype=torch.int64).chunk(world_size))
+    alltoall_cpu(rank, world_size, recv_counts, send_counts)
+
+    #allocate buffers to receive node-ids
+    recv_nodes = []
+    for i in recv_counts:
+        recv_nodes.append(torch.zeros([i.item()], dtype=torch.int64))
+
+    #form the outgoing message
+    send_nodes = []
+    for i in range(world_size):
+        send_nodes.append(torch.Tensor(global_nids_ranks[i]).type(dtype=torch.int64))
+
+    #send-recieve messages
+    alltoallv_cpu(rank, world_size, recv_nodes, send_nodes)
+
+    #TODO: This code is not needed, and will be same as the sizes of the sent global node ids
+    # in the very first exchange. Please remove this piece of code. 
+    # for each of the received orig-node-id requests lookup and send out the global node id
+    send_sizes = [len(x.tolist()) for x in recv_nodes]
+    send_counts = list(torch.Tensor(send_sizes).type(dtype=torch.int64).chunk(world_size))
+    recv_counts = list(torch.zeros([world_size], dtype=torch.int64).chunk(world_size))
+    alltoall_cpu( rank, world_size, recv_counts, send_counts)
+
+    # allocate buffers to receive global-ids
+    recv_shuffle_global_nids = []
+    for i in recv_counts:
+        recv_shuffle_global_nids.append(torch.zeros([i.item()], dtype=torch.int64))
+
+    # Use node_data to lookup global id to send over.
+    send_nodes = []
+    for proc_i_nodes in recv_nodes:
+        #list of node-ids to lookup
+        global_nids = proc_i_nodes.numpy()
+        if (len(global_nids) != 0):
+            common, ind1, ind2 = np.intersect1d(node_data[:,3], global_nids, return_indices=True)
+            values = node_data[ind1,0]
+            send_nodes.append(torch.Tensor(values).type(dtype=torch.int64))
+        else:
+            send_nodes.append(torch.Tensor(np.empty(shape=(0,))).type(dtype=torch.int64))
+
+    #send receive global-ids
+    alltoallv_cpu(rank, world_size, recv_shuffle_global_nids, send_nodes)
+
+    #form the lists with global-ids and orig-node-ids
+    #recv_shuffle_global_nids = [x.tolist() for x in recv_shuffle_global_nids]
+    #shuffle_global_nids = list(itertools.chain(*recv_shuffle_global_nids))
+    #global_nids = list(itertools.chain(*global_nids_ranks))
+    #return shuffle_global_nids, global_nids 
+
+    shuffle_global_nids = [x.numpy() for x in recv_shuffle_global_nids]
+    global_nids = [x for x in global_nids_ranks]
+    return np.column_stack((np.concatenate(global_nids), np.concatenate(shuffle_global_nids)))
+
+
+def get_shuffle_global_nids_edges(rank, world_size, edge_data, node_part_ids, node_data):
```

Comment:
why do we need this function. can we use the function above to implement this function?

Reply:
Marking this as resolved as per our code review discussion on Friday last week.